### DISCUSSION TITLE
MSAA issue when update scene graph

### DISCUSSION THREAD

**Discussion by yangcshen** [vsg-dev/VulkanSceneGraph#820] [General]
Hi,

I am learning graphics pipeline and I have the following code which modifies a pipeline state.
```
  void updatePipelineState(bool show) {
    auto renderGraph = m_RenderGraph.ref_ptr(); // m_RenderGraph is a member variable which is already defined before this function.
    if (!renderGraph)
      return;

    auto renderPass = renderGraph->getRenderPass();
    if (!renderPass)
      return;

    auto ref_selected = m_Selected.ref_ptr(); // m_Selected is a my quad from builder->createQuad()
    if (ref_selected) {
      for (auto &sc : ref_selected->stateCommands) {
        if (auto bgp = sc->cast<vsg::BindGraphicsPipeline>()) {
          for (auto &pipelineState : bgp->pipeline->pipelineStates) {
            if (auto rasterizationState =
                    pipelineState.cast<vsg::RasterizationState>()) {
              auto alternate_rasterizationState =
                  vsg::RasterizationState::create(*rasterizationState);
              alternate_rasterizationState->rasterizerDiscardEnable =
                  show ? VK_FALSE : VK_TRUE;
              pipelineState = alternate_rasterizationState;
            }
          }
        }
      }
      vsg::UpdateGraphicsPipelines updateGraphicsPipelines;

      updateGraphicsPipelines.context =
          vsg::Context::create(renderPass->device);
      updateGraphicsPipelines.context->renderPass = renderPass;

      renderGraph->accept(updateGraphicsPipelines);
    }
  }
```

I used vsg::Builder to create a quad (builder->createQuad()) and implemented ImGui button that will execute the above code. The idea is that when I click the button, it will change a pipeline state, which in the case is the `rasterizerDiscardEnable`.

Now the issue I encountered is 
If I have `windowTraits->samples = VK_SAMPLE_COUNT_1_BIT;`, this code works as expected. No problems.
But if I change windowTraits->samples to any other value, say VK_SAMPLE_COUNT_8_BIT, it all looks good until it executes the state change code above. The application will freeze and become non-responsive.

**Comment by yangcshen** [comment]
With `windowTraits->samples = VK_SAMPLE_COUNT_1_BIT;`, it works as expected.
![sample1](https://github.com/vsg-dev/VulkanSceneGraph/assets/54595761/647e7e78-5459-4fbe-b784-ef8fc1cd11c2)
But with `windowTraits->samples = VK_SAMPLE_COUNT_8_BIT;, frozen screen.
![sample8](https://github.com/vsg-dev/VulkanSceneGraph/assets/54595761/b5ebd13b-3b08-43fe-b01b-e904876ad079)

**Comment by robertosfield** [comment]
The vsgviewer example has a --samples command line option you can use to
test MSAA i.e.

    *vsgviewer models/teapot.vsgt --samples 8*

The only settings required is setting windowTraits->samples to the value
you want.  You should do this prior to creating the window, if you want to
change it at runtime then you'll need to recreate the window with new
WindowTraits, and force a rebuild of the GraphicsPipeline to honour the new
setting.
Message ID: <vsg-dev/VulkanSceneGraph/repo-discussions/820/comments/5941033@
github.com>

>

**Reply by yangcshen** [reply]
Hi Robert,

I am not changing MSAA but rasterizerDiscardEnable in vulkan. But let's say even if I change MSAA at runtime, doesn't vsg::UpdateGraphicsPipelines rebuild the GraphicsPipeline?

Thanks,
Cam

**Comment by robertosfield** [comment]
On Thu, 18 May 2023 at 21:24, Yang ***@***.***> wrote:

> I am not changing MSAA but rasterizerDiscardEnable in vulkan. But let's
> say even if I change MSAA at runtime, doesn't vsg::UpdateGraphicsPipelines
> rebuild the GraphicsPipeline?


UpdateGraphicsPipeliens does rebuild graphics pipelines ;but it won't
rebuild them unless the grphicspipline existing state is cleared by calling
release().  The vsg::Context used to provide a guideline to what settings
to inherit must be configured.

FYI, UpdateGraphicsPipeline is a very new addition so we don't have lots of
examples illustrating its use.

Another thing to consider is the use of dynamic pipeline state which allows
you to set parts of the graphics pipeline with the appropriate vulkan
commands.

Robert.

**Reply by yangcshen** [reply]
Which one is a good example for the use of dynamic pipeline state?

Thanks,
Cam

**Comment by robertosfield** [comment]
On Thu, 18 May 2023 at 21:51, Yang ***@***.***> wrote:

> Which one is a good example for the use of dynamic pipeline state?
>
There isn't one in vsgExamples. vsg::DynamicState will need configure and
assigned to your GraphicsPipeline and then in the scene graph you'll be
able to use the associated commands like vg::SetLineWidth.

Vulkan doesn't allow all graphics pipeline state to be used dynamically and
the VSG doesn't expose all the possible commands.

This is a bit of a niche feature so hasn't been tested widely.

**Reply by yangcshen** [reply]
Thanks. I will look into vsg::DynamicState. Meanwhile, I found that a quick fix for my problem was to add  
`updateGraphicsPipelines.context->overridePipelineStates.emplace_back(vsg::MultisampleState::create(renderPass->maxSamples));`

When updateGraphicsPipelines is instantiated, maxSamples is not transferred from renderPass so I had to explicitly set it. All looks good for now.
